drivers/setup/blue/blue.c

```diff
@@ -1209,15 +1456,53 @@ ScrIoControl(
             PUCHAR Src, Dest;
             UINT32 SrcDelta, DestDelta, i;
 
+            /* Validate output buffer */
+            if (stk->Parameters.DeviceIoControl.OutputBufferLength < sizeof(CONSOLE_DRAW))
+            {
+                Status = STATUS_INVALID_PARAMETER;
+                break;
+            }
+            ASSERT(Irp->MdlAddress);
+            ConsoleDraw = (PCONSOLE_DRAW)MmGetSystemAddressForMdl(Irp->MdlAddress);
+            if (ConsoleDraw == NULL)
+            {
+                Status = STATUS_INSUFFICIENT_RESOURCES;
+                break;
+            }
+
+            /* Check whether we have the size for the header plus the data area */
+            if (stk->Parameters.DeviceIoControl.OutputBufferLength <
+                    sizeof(CONSOLE_DRAW) + ConsoleDraw->SizeX * ConsoleDraw->SizeY * 2)
```

Comment:
Oh... unlike buffered I/O, direct I/O allows for TOCTOU attacks (i.e. the user could modify `ConsoleDraw` after this check but before it is used for the actual size calculation below).
So you need to make a local copy of this structure once, then only refer to the local version.

Reply:
Should be fixed now, also for IOCTL_CONSOLE_WRITE_OUTPUT_ATTRIBUTE and IOCTL_CONSOLE_WRITE_OUTPUT_CHARACTER .